models/kernels/Nonlinear_Reservoir.hpp

```diff
@@ -0,0 +1,153 @@
+class Nonlinear_Reservoir
+{
+    public:
+
+    //Constructor for a reservoir with no outlets.
+    Nonlinear_Reservoir(double minimum_storage_meters, double maximum_storage_meters, double current_storage_height_meters)
+    {
+        parameters.minimum_storage_meters = minimum_storage_meters;
+        parameters.maximum_storage_meters = maximum_storage_meters;
+        state.current_storage_height_meters = current_storage_height_meters;
+    }
+
+    //Constructor for a reservoir with only one outlet.
+    Nonlinear_Reservoir(double minimum_storage_meters, double maximum_storage_meters, double current_storage_height_meters, double a, 
+    double b, double activation_threshold_meters) : Nonlinear_Reservoir(minimum_storage_meters, maximum_storage_meters, 	current_storage_height_meters)
+    {
+        //Ensure that the activation threshold is less than the maximum storage
+        //if (activation_threshold_meters > maximum_storage_meters)
+        //TODO: Return appropriate error
+
+        this->outlets.push_back(Reservoir_Outlet(a, b, activation_threshold_meters));
+    }
+
+    //Constructor for a reservoir with multiple outlets.
+    Nonlinear_Reservoir(double minimum_storage_meters, double maximum_storage_meters, double current_storage_height_meters, std::vector <Reservoir_Outlet> &outlets) : Nonlinear_Reservoir(minimum_storage_meters, maximum_storage_meters, current_storage_height_meters)
+    {
+
+        this->outlets = outlets;
+	  
+        //Ensure that reservoir outlet activation thresholds are sorted from least to greatest height
+        sort(this->outlets.begin(), this->outlets.end(), [](const Reservoir_Outlet &left, const Reservoir_Outlet &right)
+        {              
+            return left.activation_threshold_meters < right.activation_threshold_meters;
+        });
+
+        //Ensure that the activation threshold is less than the maximum storage
+        for (auto& outlet : this->outlets) //pointer to outlets
+        {
+            if (outlet.activation_threshold_meters > maximum_storage_meters)
+            //TODO: Return appropriate error
+                continue; 
+        }
+
+    }	
+
+    //Function to update the response of storage in meters to an influx and timestep.
+    double response_storage_meters(double in_flux_meters_per_second, int delta_time_seconds)
+    {	
+        state.current_storage_height_meters += in_flux_meters_per_second * delta_time_seconds;
+	
+        for (auto& outlet : this->outlets) //pointer to outlets
+        {
+            //Update storage from outlet velocity multiplied by delta time.
+            state.current_storage_height_meters -= outlet.velocity_meters_per_second(parameters, state) * delta_time_seconds;
+        }
+
+        //If storage is less than minimum storage, correct to minimum storage.
+        if (state.current_storage_height_meters < parameters.minimum_storage_meters)
```

Comment:
Is this possible given that velocity_meters_per_second won't take water if it isn't available (i.e. storage > activation).  Also if this does happen and we do decide to correct, then we must keep track of how we corrected by, since this is arbitrary water created.  If we don't have it then we cannot properly mass balance.

Reply:
This check is after the reservoir calls the response on the vector of outlets, so it is checking the resulting storage after the response. I was trying various inputs to the unit tests and did come up with a negative storage. I don't know if the calculations should allow this to be possible though. I will go back and check through those now. I agree that the mass balance needs to be corrected from this. Fred mentioned some scenarios where it might be necessary to pass a flag out of the function in certain edge cases to go back and recalculate at a smaller timestep to account for these differences. I think you mentioned a similar type of idea previously. I think this would be a case where that would be helpful.

Reply:
After looking into this further. I found the only time that the unit tests result in a negative storage is if the coefficient 'a' is > 1. I will email Fred to double check on this and include you, Nels. We might be able to throw out this below 0 check in the code if the algorithm doesn't allow it.

Reply:
Fred says that the timestep is too long in this case. Back to the original plan then.